packages/dx-react-grid/docs/guides/custom-plugin-development.md

```diff
@@ -0,0 +1,687 @@
+          <Toolbar />
+        </Grid>
+      </Paper>
+    );
+  }
+}
+```
+*data.js:*
+```js
+const customers = [{
+    ID: 1,
+    CompanyName: 'Super Mart of the West',
+    Address: '702 SW 8th Street',
+    City: 'Bentonville',
+    State: 'Arkansas',
+    Zipcode: 72716
+  }, {
+    ID: 2,
+    CompanyName: 'Electronics Depot',
+    Address: '2455 Paces Ferry Road NW',
+    City: 'Atlanta',
+    State: 'Georgia',
+    Zipcode: 30339
+  }, {
+    ID: 3,
+    CompanyName: 'K&S Music',
+    Address: '1000 Nicllet Mall',
+    City: 'Minneapolis',
+    State: 'Minnesota',
+    Zipcode: 55403
+  }, {
+    ID: 4,
+    CompanyName: "Tom's Club",
+    Address: '999 Lake Drive',
+    City: 'Issaquah',
+    State: 'Washington',
+    Zipcode: 98027
+  }, {
+    ID: 5,
+    CompanyName: 'E-Mart',
+    Address: '3333 Beverly Rd',
+    City: 'Hoffman Estates',
+    State: 'Illinois',
+    Zipcode: 60179
+  }];
+
+export default {
+  getCustomers() {
+    return customers;
+  }
+}
+```
+
+Let's take a look at the screenshot below to see the final result that we will get after performing all the steps from the tutorial:
+
+![](../../../../../img/guides/custom-plugin-development/result.png)
+
+As you can see, there is a select box at the top right corner of the grid associated with the `CompanyName` column. When a user selects an item from the box, the grid displays only filtered items. Moreover, there is the `CLEAR` button that clears the selected value. When the value is cleared, the grid displays all available rows.
+
+
+Let's start extending our grid!
+
+## Rendering Custom Markup
+For rendering custom markup in existing grid UI elements, in our case it is the toolbar, we need to create a custom plugin. For creating a plugin, we need to wrap all content in the [Plugin](../../../core/docs/reference/plugin.md) component. Since our plugin will reside in the grid's toolbar, we will name it `ToolbarFilter`. Let's create a new folder `plugins` in our project and
+put a new `toolbar-filter.js` file into the folder with the following code:
+
+*toolbar-filter.js:*
+```jsx
+import {
+  Plugin
+} from '@devexpress/dx-react-core';
+
+export class ToolbarFilter extends React.PureComponent {
+  render() {
+    return (
+      <Plugin
+        name="ToolbarFilter"
+      >
+      </Plugin>
+    );
+  }
+}
+```
+We are going to show our plugin in the Grid's [toolbar](../reference/toolbar.md). According to the documentation, the toolbar exports the `toolbarContent` template. So, we need to use this template in our plugin for modifying its content. For this task we need to add the [Template](../../../core/docs/reference/template.md) component to our plugin and set its `name` property to `toolbarContent`.
+```jsx
+...
+import {
+  ...,
+  Template
+} from '@devexpress/dx-react-core';
+
+export class ToolbarFilter extends React.PureComponent {
+  render() {
+    return (
+      <Plugin
+        ...
+      >
+        <Template name="toolbarContent">
+          </Template>
```

Comment:
Unnecessary spaces.

Reply:
Can be just <Template name="toolbarContent" **/**>